posydon/binary_evol/MESA/step_mesa.py

```diff
@@ -1251,29 +1259,39 @@ def __call__(self, binary):
         state_1 = self.binary.star_1.state
         state_2 = self.binary.star_2.state
         event = self.binary.event
-        mass_ratio = self.binary.star_2.mass/self.binary.star_1.mass
+        m1 = self.binary.star_1.mass
+        m2 = self.binary.star_2.mass
+        mass_ratio = m2/m1
         p = self.binary.orbital_period
-        p_max_HMS_grid = 6500
-        if (state_1 == 'H-rich_Core_H_burning'
-                and state_2 == 'H-rich_Core_H_burning' and event == 'ZAMS'
-                and mass_ratio <= 1. and p <= p_max_HMS_grid):
+        if (state_1 == 'H-rich_Core_H_burning' and 
+            state_2 == 'H-rich_Core_H_burning' and 
+            event == 'ZAMS' and
+            self.m1_min <= m1 <= self.m1_max and 
+            self.q_min <= mass_ratio <= self.q_max and 
+            self.p_min <= p <= self.p_max):
             self.flip_stars_before_step = False
             super().__call__(self.binary)
-        elif (state_1 == 'H-rich_Core_H_burning'
-              and state_2 == 'H-rich_Core_H_burning'
-              and event == 'ZAMS' and mass_ratio > 1. and p <= p_max_HMS_grid):
+        elif (state_1 == 'H-rich_Core_H_burning' and 
+              state_2 == 'H-rich_Core_H_burning' and 
+              event == 'ZAMS' and
+              self.m1_min <= m2 <= self.m1_max and
+              1./self.q_min >= mass_ratio > self.q_max and 
+              self.p_min <= p <= self.p_max):
             self.flip_stars_before_step = True
             super().__call__(self.binary)
-        elif (state_1 == 'H-rich_Core_H_burning'
-              and state_2 == 'H-rich_Core_H_burning'
-              and event == 'ZAMS'
-              and p > p_max_HMS_grid):              # redirect if outside grid
+        # redirect if outside grid
+        elif (state_1 == 'H-rich_Core_H_burning' and 
+              state_2 == 'H-rich_Core_H_burning' and 
+              event == 'ZAMS' and 
+              p > self.p_max):
             self.binary.event = 'redirect'
             return
-        elif (state_1 == 'H-rich_Central_C_depletion'):     # redirect if CC1
+        # redirect if CC1
+        elif (state_1 == 'H-rich_Central_C_depletion'):     
```

Comment:
Just a question about the stars, where we stop the evolution before carbon depletion: do they still get to the stellar state 'H-rich_Central_C_depletion'?

Reply:
No, they will not. They will likely get `H-rich_Central_He_depleted` as we no longer have a state `near_C_depletion`. But then because we know the binary system reached the end of the MESA evolution successfully (no `initial_MT` or `unstable_MT`), we update the binary event to `CC1` or `CC2`, and the `step_SN` will collapse such binary (see `STAR_STATES_CC` in `binary_evol/flow_chart.py`). I vaguely remember why this redirect condition was introduced. How do we hit HMS-HMS grids while one of the stars was previously evolved? This redirection might be obsolete.

Reply:
@devimisra @ezapartas do you remember why we introduced this redirection in the HMS-HMS step? It had to do with q=1 binaries in v1 reaching two CC1 one after the other. But a `CO--H_rich_Central_He_depleted` binary would not reach this condition, it will be redirected by `step_detached` to `CC2`.